### ISSUE TITLE
document how intellij bazel rules depend on scala rules

### ISSUE THREAD

**Issue by johnynek** [bazelbuild/rules_scala#466]
in this file there are two references to intellij:

https://github.com/bazelbuild/rules_scala/blob/master/scala/scala.bzl#L774

It seems that intellij seems to be wanting some information put in the struct accessible at `.scala` on a target. However, we want to move to modern providers, so that seems to mean that the intellij rules will need to be updated.

Can we document:

1. what dependency do the intellij rules have on us? What API exactly do they require?
2. what we need to do to change the API. I guess we will need to make changes in almost lockstep: so we make the breaking change here then send a PR to some other repo.

@ittaiz @gkk-stripe @andyscott do you all know anything about this?

**Comment by andyscott** [comment]
I don't have any context on this-- but I am also extremely curious to learn more.

**Comment by johnynek** [comment]
cc @jjudd who also has a "blame" on one of the lines that mentions intellij (although ittai seems to have the most).

**Comment by gkossakowski** [comment]
i believe the dependency comes from intellij collecting java provider and expecting certain attributes on it, see: https://github.com/bazelbuild/intellij/blob/master/aspect/intellij_info_impl.bzl#L408 and calls to that method

**Comment by johnynek** [comment]
yeah, I wonder why they do that rather than just getting the java provider? I guess we would need a PR to them.

@ittaiz you have followed this: does the native java provider have enough information to support just using the java provider and not having any special rules for scala?

**Comment by ittaiz** [comment]
The important part is a few lines below what @gkossakowski linked to:
```
  if hasattr(target, "scala"):
    return target.scala
  if hasattr(target, "kt") and hasattr(target.kt,"outputs"):
    return target.kt
  # java_common.provider is a work in progress. It will soon expose the information
  # we require (e.g. outputs jars, jdeps), but does not yet do so.
  if java_common.provider in target:
    return target[java_common.provider]
```
Note:

>   java_common.provider is a work in progress. It will soon expose the information
>   we require (e.g. outputs jars, jdeps), but does not yet do so.

I think the work @dbabkin did on the [bazel side](https://github.com/bazelbuild/bazel/issues/3769) took us a lot farther but unfortunately due to some ongoing [discussion](https://github.com/bazelbuild/bazel/issues/4584) with respect to strict-deps labels I'm not sure we can switch to using the new JavaInfo API (I think, not sure, it hides information we currently need for the labels accumulation).

The bottom line unless I'm mistaken is that as soon as we'll be able to finalize the labels issue we'll move to the new API (which I think is not yet released).

cc @chaoren @hmemcpy for people who have intimate knowledge of that area
